# Примеры преобразования размерных величин 

In [1]:
import sympy as sp
import scipy.constants as const

## Размерности некоторых величин применяемых в расчетах

Преобразования размерностей ключевых величин полезно знать наизусть. Однако всегда можно найти их в таблицах. Полезно знать, что многие физические константы зашины в модуле scipy.constants.
Рассмотрим размерности ряда величин широко применяемых в нефтяном инжиниринге.

In [2]:
# в модуле scipy.constants есть значения общепринятых констант
print(const.pi)

3.141592653589793


### Объемный расход $q$

В СИ измеряется в [м$^3$/сек], в практических метрических единицах измеряется в [м$^3$/сут], в американских промысловых единицах измеряется в [bbl/day].

* $1$ [м$^3$/сек] = $543439$ [bbl/day] = $86400$ [м$^3$/сут]
* $1$ [м$^3$/сут] = $\dfrac{1}{86400}$ [м$^3$/сек] $= 1.157407 \cdot 10^{-5} $ [м$^3$/сек]
* $1$ [bbl/day] = $0.15898$ [м$^3$/сут]

In [3]:
# полезно помнить 
print(f'Количество секунд в сутках = {24*60*60} = {const.day}')
print(f'Один  [м3/сут] в  [м3/сек] = {1/const.day}')
print(f'Один баррель в день [bbl/day] в метрах кубических в день [м3/сут] = {const.bbl}')
print(f'Один баррель в день [bbl/day] в метрах кубических в секунду [м3/сут] = {const.bbl/const.day}')
print(f'Один метр кубический в день [м3/сут] в метрах кубических в секунду [м3/сут] = {1/const.day}')
 


Количество секунд в сутках = 86400 = 86400.0
Один  [м3/сут] в  [м3/сек] = 1.1574074074074073e-05
Один баррель в день [bbl/day] в метрах кубических в день [м3/сут] = 0.15898729492799998
Один баррель в день [bbl/day] в метрах кубических в секунду [м3/сут] = 1.8401307283333331e-06
Один метр кубический в день [м3/сут] в метрах кубических в секунду [м3/сут] = 1.1574074074074073e-05


### Проницаемость $k$

В СИ измеряется в [м$^2$], в практических метрических единицах измеряется в [мД], в американских промысловых единицах измеряется в [mD].

Определение: в пористой среде с проницаемостью в один Дарси для поддержания течения жидкости с динамической вязкостью 1 сП со скоростью фильтрации 1 см/с необходимо поддерживать перепад давления жидкости приблизительно в одну атмосферу на 1 см вдоль направления течения. При использовании физической атмосферы для расчета перепада давления (физическая атмосфера = 101 325 Па) единица проницаемости равняется приблизительно 0.986923 мкм². 

В отечественной литературе при определении дарси в качестве величины атмосферы было принято использовать техническую атмосферу (1 кгс/см² = 98 066,5 Па), так что для величины дарси получалось значение приблизительно 1,02 мкм², причём эпизодические случаи использования западного определения дарси специально отмечались [[ru.wikipedia.org/wiki/Дарси](https://ru.wikipedia.org/wiki/%D0%94%D0%B0%D1%80%D1%81%D0%B8)]. Согласно  ГОСТ 26450.2-85 величины 1 Дарси $= 0.9869⋅10^{−12}$ м².

* $1$ [м$^2$] = $1.01325 \cdot 10^{15}$ [мД]
* $1$ [мД] = $0.986923 \cdot 10^{-15}$ [м$^2$]

In [4]:
print(f'Один миллиДарси [мД] в метрах квадратных [м²] = {1e5/const.atm * 1e-15}')

Один миллиДарси [мД] в метрах квадратных [м²] = 9.86923266716013e-16




### Вязкость $\mu$

* $1$ [Па$\cdot$с] = $1000$ [сП]
* $1$ [сП] = $10^{-3}$ [Па$\cdot$с]

### Давление $p$
* $1$ [Па] = $0.0001450$ [psi] = $0.00000987$ [атм]
* $1$ [атм] = $14.6959$ [psi] = $101325$ [Па]


In [5]:
AT = 98066.5  # technical atmosphere in Pa,  техническая атмосфера в Па
print(f'Один  [psi] в [Па] = {const.psi}')
print(f'Один  [bar] в [Па] = {const.bar}')
print(f'Один  [atm] в [Па] = {const.atm}')
print(f'Один  [at] в [Па] = {AT}')
print(f'Один  [atm] в [psi] = {const.atm/const.psi}')

Один  [psi] в [Па] = 6894.757293168361
Один  [bar] в [Па] = 100000.0
Один  [atm] в [Па] = 101325.0
Один  [at] в [Па] = 98066.5
Один  [atm] в [psi] = 14.69594877551345



### Расстояние $x$

*$1$ [м] = $3.28$ [ft]



# Размерный коэффициент для формулы Дюпюи


$$ Q = \frac{ 2 \pi kh}{ \mu B} \frac{  \left( p_i - p \right) } {\ln{\dfrac{r_e}{r_w}} +S } $$



In [6]:
# зададим переменные sympy
Q, k, h, mu, B, pres, pwf, re, rw, S, pi = sp.symbols('Q k h mu B p_res p_wf r_e r_w S pi', real=True, positive=True)

In [7]:
# определим уравнение
eq = sp.Eq(Q, 2 * pi * k * h / (mu * B) * (pres - pwf) / (sp.ln(re/rw) + S))
eq

Eq(Q, 2*h*k*pi*(p_res - p_wf)/(B*mu*(S + log(r_e/r_w))))

In [8]:
# подставим в уравнение переводные величины
eq = eq.subs(Q, 1/const.day * Q)  # дебит, коэффициент равен одному [м3/сут] в [м3/сек]
f_k = 1e5/const.atm * 1e-15
print(f'Один миллиДарси [мД] в метрах квадратных [м²] = {f_k}')
eq = eq.subs(k, f_k * k)     # проницаемость: Один миллиДарси [мД] в метрах квадратных [м²]
eq = eq.subs(mu, 1e-3 * mu)
eq = eq.subs(pres, 101325 * pres)
eq = eq.subs(pwf, 101325 * pwf)
eq = eq.subs(pi, 3.1416)
#eq = eq.subs(rw, 3.28*rw)
#eq = eq.subs(h, 3.28*h)
display(eq)
print('решим относительно Q и упростим')
eq1 = sp.simplify(sp.solve(eq,Q)[0])
display(sp.Eq(Q,eq1))

Один миллиДарси [мД] в метрах квадратных [м²] = 9.86923266716013e-16


Eq(1.15740740740741e-5*Q, 6.20103626943005e-12*h*k*(101325*p_res - 101325*p_wf)/(B*mu*(S + log(r_e/r_w))))

решим относительно Q и упростим


Eq(Q, 0.054286848*h*k*(p_res - p_wf)/(B*mu*(S + log(r_e/r_w))))

In [9]:
f = 1/eq1.args[0]
f

18.4206679304719

In [10]:
a = sp.symbols('a')
eq2 = eq1.subs(eq1.args[0],1/a)
with sp.evaluate(False):
    display(eq2.subs(a, f))

h*k*(p_res - p_wf)/(18.4206679304719*B*mu*(S + log(r_e/r_w)))


$$ Q = \frac{ kh}{18.42  \mu B} \frac{  \left( p_i - p \right) } {\ln{\dfrac{r_e}{r_w}} +S } $$
